File: js-pr/56/javascript/code-quality-readability-maintainability/cache_manager_44.js
// {fact rule=code-quality-readability-maintainability@v1.0 defects=0}
class LRUCache {
    constructor(maxSize = 100) {
        this.cache = new Map();
        this.accessCounts = new Map();
        this.maxSize = maxSize;
    }
    
    get(key) {
        if (this.cache.has(key)) {
            this.updateAccessCount(key);
            return this.cache.get(key);
        }
        return null;
    }
    
    set(key, value) {
        if (this.cache.size >= this.maxSize) {
            this.evictLeastUsedItem();
        }
        
        this.cache.set(key, value);
        this.accessCounts.set(key, 1);
    }
    
    updateAccessCount(key) {
        const currentCount = this.accessCounts.get(key) || 0;
        this.accessCounts.set(key, currentCount + 1);
    }
    
    evictLeastUsedItem() {
        const leastUsedKey = this.findLeastUsedKey();
        this.cache.delete(leastUsedKey);
        this.accessCounts.delete(leastUsedKey);
    }
    
    findLeastUsedKey() {
        let minCount = Infinity;
        let leastUsedKey = null;
        
        for (const [key, count] of this.accessCounts.entries()) {
            if (count < minCount) {
                minCount = count;
                leastUsedKey = key;
            }
        }
        
        return leastUsedKey;
    }
}
// {/fact}

File: python-pr/52/python/code-quality-readability-maintainability/cache_manager_44.py
# {fact rule=code-quality-readability-maintainability@v1.0 defects=0}
class LRUCache:
    def __init__(self, max_size=100):
        self.cache = {}
        self.access_counts = {}
        self.max_size = max_size
    
    def get(self, key):
        if key in self.cache:
            self._update_access_count(key)
            return self.cache[key]
        return None
    
    def set(self, key, value):
        if len(self.cache) >= self.max_size:
            self._evict_least_used_item()
        
        self.cache[key] = value
        self.access_counts[key] = 1
    
    def _update_access_count(self, key):
        self.access_counts[key] = self.access_counts.get(key, 0) + 1
    
    def _evict_least_used_item(self):
        least_used_key = min(self.access_counts, key=self.access_counts.get)
        del self.cache[least_used_key]
        del self.access_counts[least_used_key]
# {/fact}